presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControl.java

```diff
@@ -0,0 +1,328 @@
+import static com.facebook.presto.spi.security.AccessDeniedException.denyDeleteTable;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyDropSchema;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyDropTable;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyDropView;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyGrantTablePrivilege;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyInsertTable;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyRenameColumn;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyRenameSchema;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyRenameTable;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyRevokeTablePrivilege;
+import static com.facebook.presto.spi.security.AccessDeniedException.denySelectTable;
+import static com.facebook.presto.spi.security.AccessDeniedException.denySelectView;
+import static com.facebook.presto.spi.security.AccessDeniedException.denySetCatalogSessionProperty;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyShowSchemas;
+import static com.facebook.presto.spi.security.AccessDeniedException.denyShowTables;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static java.util.Objects.requireNonNull;
+
+public class FileBasedSystemAccessControl
+        implements SystemAccessControl
+{
+    private static final String CATALOG_ACCESS_DENIED = "Access denied for catalog ";
+    public static final String NAME = "file";
+
+    private static final FileBasedSystemAccessControl INSTANCE = new FileBasedSystemAccessControl(jsonCodec(FileBasedSystemAccessControlRules.class));
+
+    private final JsonCodec<FileBasedSystemAccessControlRules> rulesJsonCodec;
+    @JsonProperty("catalogs") private List<CatalogAccessControlRule> catalogRules;
+
+    @Inject
+    private FileBasedSystemAccessControl(JsonCodec<FileBasedSystemAccessControlRules> rulesCodec)
+    {
+        rulesJsonCodec = rulesCodec;
+    }
+
+    public static class Factory
+            implements SystemAccessControlFactory
+    {
+        @Override
+        public String getName()
+        {
+            return NAME;
+        }
+
+        @Override
+        public SystemAccessControl create(Map<String, String> config)
+        {
+            final String CONFIG_FILE_NAME = "security.config-file";
+            requireNonNull(config, "config is null");
+
+            try {
+                Path path = Paths.get(config.get(CONFIG_FILE_NAME));
+                if (!path.isAbsolute()) {
+                    path = path.toAbsolutePath();
+                }
+                FileBasedSystemAccessControlRules rules = INSTANCE.rulesJsonCodec.fromJson(Files.readAllBytes(path));
+                INSTANCE.catalogRules = rules.getCatalogRules();
```

Comment:
Changing a private field here is not a good idea. It'd be better to just create the SystemAccessControl here in the factory; we're not planning to make a ton of these, so it should be fine. And then you can move the rule parsing into the constructor.